src/auto/pixbuf.rs

```diff
@@ -75,90 +76,98 @@ impl Pixbuf {
-        {
-            callback_guard!();
-            unsafe {
-                let mut error = ptr::null_mut();
-                let mut width = mem::uninitialized();
-                let mut height = mem::uninitialized();
-                let _ = ffi::gdk_pixbuf_get_file_info_finish(res, &mut width, &mut height, &mut error);
-                let result = if error.is_null() { Ok((width, height)) } else { Err(from_glib_full(error)) };
-                let callback: &&(Fn(Result<(i32, i32), Error>) + Send + Sync + 'static) = transmute(user_data);
-                callback(result);
-            }
-        }
-        let callback = get_file_info_async_trampoline;
+    pub fn new_from_xpm_data(data: &[&str]) -> Pixbuf {
         unsafe {
-            ffi::gdk_pixbuf_get_file_info_async(filename.to_glib_none().0, cancellable.0, Some(callback), Box::into_raw(user_data) as *mut _);
+            from_glib_full(ffi::gdk_pixbuf_new_from_xpm_data(data.to_glib_none().0))
         }
     }
 
+    //pub fn from_pixdata(pixdata: /*Ignored*/&Pixdata, copy_pixels: bool) -> Result<Pixbuf, Error> {
+    //    unsafe { TODO: call ffi::gdk_pixbuf_from_pixdata() }
+    //}
+
     pub fn get_formats() -> Vec<PixbufFormat> {
         unsafe {
             FromGlibPtrContainer::from_glib_container(ffi::gdk_pixbuf_get_formats())
         }
     }
-
-    //pub fn new_from_stream_async<'a, P: IsA</*Ignored*/gio::InputStream>, Q: Into<Option<&'a gio::Cancellable>>, R: Fn(Result<(), Error>) + Send + Sync + 'static>(stream: &P, cancellable: Q, callback: R) {
-    //    unsafe { TODO: call ffi::gdk_pixbuf_new_from_stream_async() }
-    //}
-
-    //pub fn new_from_stream_at_scale_async<'a, P: IsA</*Ignored*/gio::InputStream>, Q: Into<Option<&'a gio::Cancellable>>, R: /*Ignored*/gio::AsyncReadyCallback>(stream: &P, width: i32, height: i32, preserve_aspect_ratio: bool, cancellable: Q, callback: R) {
-    //    unsafe { TODO: call ffi::gdk_pixbuf_new_from_stream_at_scale_async() }
-    //}
 }
 
 pub trait PixbufExt {
+    fn add_alpha(&self, substitute_color: bool, r: u8, g: u8, b: u8) -> Option<Pixbuf>;
+
     fn apply_embedded_orientation(&self) -> Option<Pixbuf>;
 
+    fn composite(&self, dest: &Pixbuf, dest_x: i32, dest_y: i32, dest_width: i32, dest_height: i32, offset_x: f64, offset_y: f64, scale_x: f64, scale_y: f64, interp_type: InterpType, overall_alpha: i32);
+
+    fn composite_color(&self, dest: &Pixbuf, dest_x: i32, dest_y: i32, dest_width: i32, dest_height: i32, offset_x: f64, offset_y: f64, scale_x: f64, scale_y: f64, interp_type: InterpType, overall_alpha: i32, check_x: i32, check_y: i32, check_size: i32, color1: u32, color2: u32);
+
+    fn composite_color_simple(&self, dest_width: i32, dest_height: i32, interp_type: InterpType, overall_alpha: i32, check_size: i32, color1: u32, color2: u32) -> Option<Pixbuf>;
+
+    fn copy(&self) -> Option<Pixbuf>;
```

Comment:
Can this really fail? And the others returning an Option

Reply:
Bad, but yes: https://developer.gnome.org/gdk-pixbuf/stable/gdk-pixbuf-Image-Data-in-Memory.html#gdk-pixbuf-copy

Reply:
I will check all others later.

Reply:
Interesting